CMakeLists.txt

```diff
@@ -73,7 +73,7 @@ if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" OR
     set(CMAKE_CXX_FLAGS_RELEASE "-O2")
     set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "-O2 -g")
     #   Enable lots of warnings
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wpedantic -Werror -Wno-deprecated-declarations -Wswitch-enum")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wpedantic -Werror -Wno-deprecated-declarations -Wno-return-type -Wswitch-enum")
```

Comment:
I'm not sure if this change is _correct_ - the commit message indicates that this would trigger on code marked with our internal `UNREACHABLE` macro (unless I've misunderstood it), but these are _supposed_ to behave in the way they do to indicate a path that the developer intended to be unreachable (and fail loudly if they are not).

Reply:
The problem I get when compiling is sometimes there the function ends in a `UNREACHABLE` macro but not a return statement. Since GCC does not know that this is unreachable, it complains that there is a path in the non-void function that does not end in a `return` statement. Hence, it throws a compiler warning, an since `-Wall` is set, this warning is actually an error and compilation fails.

Reply:
I removed the last commit. Hope everything works out now.

Reply:
When building CBMC in my IDE (clion) I get the following error:

```
/usr/bin/c++ -DHAVE_MINISAT2 -DSATCHECK_MINISAT2 -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/cbmc/cmake-build-debug/src -I/home/cbmc/src -I/home/cbmc/cmake-build-debug/src/solvers -I/home/cbmc/src/solvers -I/home/cbmc/cmake-build-debug/minisat2-src -I/home/cbmc/cmake-build-debug/src/util -I/home/cbmc/src/util -I/home/cbmc/cmake-build-debug/src/big-int -I/home/cbmc/src/big-int -I/home/cbmc/cmake-build-debug/src/langapi -I/home/cbmc/src/langapi -Wall -Wpedantic -Werror -Wno-deprecated-declarations -Wswitch-enum -g -fdiagnostics-color=always -std=gnu++11 -MD -MT src/solvers/CMakeFiles/solvers.dir/flattening/boolbv_get.cpp.o -MF src/solvers/CMakeFiles/solvers.dir/flattening/boolbv_get.cpp.o.d -o src/solvers/CMakeFiles/solvers.dir/flattening/boolbv_get.cpp.o -c /home/cbmc/src/solvers/flattening/boolbv_get.cpp
/home/cbmc/src/solvers/flattening/boolbv_get.cpp: In member function ‘virtual exprt boolbvt::bv_get_rec(const exprt&, const bvt&, std::size_t) const’:
/home/cbmc/src/solvers/flattening/boolbv_get.cpp:272:1: error: control reaches end of non-void function [-Werror=return-type]
  272 | }
      | ^
cc1plus: all warnings being treated as errors
```

Reply:
Alright, here is how I reproduce the issues with gcc 12.2.1:

```
git clone git@github.com:diffblue/cbmc.git
cd cbmc
git submodule update --init
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_MAKE_PROGRAM=/path/to/ninja -G Ninja -S /home/cbmc -B /home/cbmc/build
cmake --build . --target all -j 6

```

If you are unable to reproduce it I can create a docker container

Reply:
That is helpful. Can you also confirm which OS flavour / version you have observed this issue with?

Reply:
Manjaro Linux. Kernel 6.2.9, worth noting that I had that issue some time ago as well but back then I did not end up using cbmc for my project so I never went looking. Also, I just looked and it seems to not be an issue with ninja, as I also get it when using make instead of ninja.

Reply:
That's a good suggestion, but changing it did not solve the error. I get the same error with the following compilation command: 
`/usr/bin/c++ -DHAVE_MINISAT2 -DSATCHECK_MINISAT2 -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/julian/Documents/cbmc/cmake-build-debug/src -I/home/julian/Documents/cbmc/src -I/home/julian/Documents/cbmc/cmake-build-debug/src/solvers -I/home/julian/Documents/cbmc/src/solvers -I/home/julian/Documents/cbmc/cmake-build-debug/minisat2-src -I/home/julian/Documents/cbmc/cmake-build-debug/src/util -I/home/julian/Documents/cbmc/src/util -I/home/julian/Documents/cbmc/cmake-build-debug/src/big-int -I/home/julian/Documents/cbmc/src/big-int -I/home/julian/Documents/cbmc/cmake-build-debug/src/langapi -I/home/julian/Documents/cbmc/src/langapi -Wall -Wpedantic -Werror -Wno-deprecated-declarations -Wswitch-enum -g -fdiagnostics-color=always -std=c++11 -MD -MT src/solvers/CMakeFiles/solvers.dir/flattening/boolbv_get.cpp.o -MF src/solvers/CMakeFiles/solvers.dir/flattening/boolbv_get.cpp.o.d -o src/solvers/CMakeFiles/solvers.dir/flattening/boolbv_get.cpp.o -c /home/julian/Documents/cbmc/src/solvers/flattening/boolbv_get.cpp
`